### PR TITLE
Feature/optimize parser

### PR DESCRIPTION
**Pull request by o-tsaruk**
Optimized several parser functions by using regular expressions.
Fixed error with struct field comments.

### CODE REVIEW

File: src/parser.rs [hunk 1]
```diff
@@ -821,39 +790,29 @@ impl<'a> Parser<'a> {
...
-        let event_raw = read_until(self.chars, vec![SEMICOLON])
-            .trim()
-            .replace("( ", "(")
-            .replace(" )", ")");
-
-        let tokens = split(&event_raw, " ", None);
-        let mut args_reader = ArgsReader::ArgName;
-        let mut indexed = false;
+        let event_raw = read_until(self.chars, vec![SEMICOLON]);
+        let regex = Regex::new(r#"(?P<name>^[A-Za-z0-9]+)\(\s*(?P<parameters>.*)\)"#).unwrap();
```
Comment (coreggon11) [inline src/parser.rs]: what if we have event Some_Event

File: src/parser.rs [hunk 2]
```diff
@@ -870,19 +829,16 @@ impl<'a> Parser<'a> {
...
-            if token == "{" {
-                continue
-            } else {
-                token.remove_matches(",");
-                values.push(token);
-            }
-        }
+
+        let regex =
+            Regex::new(r#"(?P<name>^[A-Za-z0-9]+)\s*\{\s*(?P<values>([A-Za-z0-9]+,?\s*)*)"#)
```
Comment (coreggon11) [inline src/parser.rs]: add _ to the chars

File: src/parser.rs [hunk 3]
```diff
@@ -896,19 +852,51 @@ impl<'a> Parser<'a> {
...
+            if ch == CURLY_CLOSE {
+                break
+            }
+            buffer.push(ch);
+        }
+        let struct_raw = buffer.replace(" => ", "=>");
+
+        let regex =
+            Regex::new(
+                r#"(?P<comment1>(\n\s*//.*)*|(\n\s*/\*(.*\n)*?.*\*/\s*))?(?P<field>\n\s*[A-Za-z0-9=>()]+\s+[A-Za-z0-9]+\s*;)(?P<comment2>(.*//.*)|(.*/\*(.*\n)*?.*\*/))?"#)
```
Comment (coreggon11) [inline src/parser.rs]: can you split it to multiple lines?
also think it would be good to apply this to enum and event

File: src/parser.rs [hunk 4]
```diff
@@ -918,17 +906,26 @@ impl<'a> Parser<'a> {
...
-        let field_type = self.convert_variable_type(tokens[0].to_owned());
-        let mut name = tokens[1].to_owned();
-        name.remove_matches(";");
-        StructField { name, field_type }
+    /// returns the comment as String
+    fn parse_struct_field_comment(&mut self, mut comment_raw: String) -> String {
+        if comment_raw.is_empty() {
+            comment_raw
+        } else if comment_raw.contains("/*") {
+            comment_raw.remove_matches("/*");
```
Comment (coreggon11) [inline src/parser.rs:997]: we want to get rid of remove matches, you can add those characters to the regex

**Review by coreggon11 — CHANGES_REQUESTED** [review]
good job, left a few comments

File: src/parser.rs (hunk 4 above)
Comment (o-tsaruk) [reply src/parser.rs:997]: if we add it to regex, we will need to 2 comments groups between field and 2 groups after. For instance:
`"(?P<comment1>(\n\s*//.*)*|(\n\s*/\*(.*\n)*?.*\*/\s*))? ..."`
we can move /* before group start, but in this case it won't work with one line comments group. As result we need to create separate one. It equal to add one more function call in parser code. Therefore I think will be better to leave it as it is. 
What do you think?

Comment (coreggon11) [reply src/parser.rs:997]: ok leave it

**Review by coreggon11 — APPROVED** [review]

**Review by coreggon11 — APPROVED** [review]

**Pull request merged**